spring-cloud-gcp-starters/spring-cloud-gcp-starter-sql/src/main/java/org/springframework/cloud/gcp/sql/CloudSqlJdbcUrlProvider.java

```diff
@@ -0,0 +1,29 @@
+/*
+ *  Copyright 2017 original author or authors.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *       http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.springframework.cloud.gcp.sql;
+
+/**
+ * Implementations of this interface typically construct a JDBC URL for Cloud SQL from a GCP
+ * project ID and an instance name.
+ *
+ * @author João André Martins
+ */
+@FunctionalInterface
+public interface CloudSqlJdbcUrlProvider {
```

Comment:
The motivation for this was to split the `DataSource` `@Bean` method in two -- constructing the JDBC URL and creating a `DataSource` with it. We wanted to give users the flexibility of using their own `DataSources` but using our URL builder.

This interface is a way to avoid a `String` `@Bean`.